### PR TITLE
Simplify storage.Interface by merging Watch and WatchList functions.

### PR DESCRIPTION
**Pull request by wojtek-t**
```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig-api-machinery

### CODE REVIEW

**Comment by wojtek-t** [comment]
/assign @jpbetz 

@jpbetz - could you help with reviewing this one?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -1239,23 +1239,19 @@ func (e *Store) Watch(ctx context.Context, options *metainternalversion.ListOpti
...
-			if err != nil {
-				return nil, err
-			}
-			if e.Decorator != nil {
-				return newDecoratedWatcher(ctx, w, e.Decorator), nil
-			}
-			return w, nil
+		if k, err := e.KeyFunc(ctx, name); err == nil {
+			key = k
+			storageOpts.Recursive = false
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1248]: Thanks for cleaning this up while adding the recursive support.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 2]
```diff
@@ -1239,23 +1239,19 @@ func (e *Store) Watch(ctx context.Context, options *metainternalversion.ListOpti
 
 // WatchPredicate starts a watch for the items that matches.
 func (e *Store) WatchPredicate(ctx context.Context, p storage.SelectionPredicate, resourceVersion string) (watch.Interface, error) {
-	storageOpts := storage.ListOptions{ResourceVersion: resourceVersion, Predicate: p}
+	storageOpts := storage.ListOptions{ResourceVersion: resourceVersion, Predicate: p, Recursive: true}
+
+	key := e.KeyRootFunc(ctx)
 	if name, ok := p.MatchesSingle(); ok {
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1245]: If `p.MatchesSingle()` is true but `e.KeyFunc(ctx, name)` returns an error, should `storageOpts.Recursive` still be set to `false`?

**Review by jpbetz — COMMENTED** [review]
Just one question. LGTM otherwise.

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1245]: Yeah - this is consistent with what we had before, where we were calling WatchList with KeyRootFunc below.

**Pull request merged**